### PR TITLE
Add proposal for pattern matching Span<char> on a constant string

### PR DESCRIPTION
**Pull request by YairHalberstadt**

### CODE REVIEW

File: proposals/csharp-10.0/pattern-match-span-of-char-on-string.md
Comment (RikkiGibson) [inline proposals/csharp-10.0/pattern-match-span-of-char-on-string.md:19]: preffered->preferred :)

**Comment by jaredpar** [comment]
Two meta points about this PR that I wanted to get feedback on from @333fred and @jcouv 

- This PR is dropping a proposal directly into `proposals/csharplang-10.0`. In the past we dropped them into `proposals` and then migrated to the version specific folder close to release. In this particular case though the feature is approved, there is a PR to push it into the language so the intermediate step seems unneeded. Are we fine with this approach? 
- The PR is modifying an existing proposal in line. So if you go to the csharp-7.0 folder you now see the 7.0 version of pattern matching with a mod to say "here is how it changed in C# 10". I'm still weighing if this is the best approach or not. For span safety lifetime I'd been thinking of taking the span safety doc, copying it into the C# 10 folder and then changing it as needed. That way it left the C# 7.2 spec in the original form. It would be hard to change that spec inline and add C# 10 specific sections and keep the overall spec readable. Think we should figure out what overall approach we want to lean towards here.

File: proposals/csharp-10.0/pattern-match-span-of-char-on-string.md
Comment (jaredpar) [reply proposals/csharp-10.0/pattern-match-span-of-char-on-string.md:37]: Need to also discuss how both the `AsSpan` and `SequencEquals` methods are chosen. Think it's easiest here to say that we are looking for a signature match (don't consider signatures where there are optional members). 

This is a frequent source of discussion in compiler and LDM. Best to get it set from the start here. The records spec has a lot of similar discussion that you can draw from.

File: proposals/csharp-10.0/pattern-match-span-of-char-on-string.md
Comment (jaredpar) [inline proposals/csharp-10.0/pattern-match-span-of-char-on-string.md:30]: I would either end with `, etc ...` to make it clear there are missing entries here or take the time to fill them all out. One example I see missing is `nameof()`.

File: proposals/csharp-10.0/pattern-match-span-of-char-on-string.md
Comment (YairHalberstadt) [reply proposals/csharp-10.0/pattern-match-span-of-char-on-string.md:23]: I would want to leave this up to the compiler, as there's definitely room for improvement with the current optimizations (in some cases it can be more efficient to emit a trie for example).

File: proposals/csharp-10.0/pattern-match-span-of-char-on-string.md
Comment (YairHalberstadt) [reply proposals/csharp-10.0/pattern-match-span-of-char-on-string.md:37]: Changes made

File: proposals/csharp-10.0/pattern-match-span-of-char-on-string.md
Comment (YairHalberstadt) [reply proposals/csharp-10.0/pattern-match-span-of-char-on-string.md]: Done

**Comment by 333fred** [comment]
> This PR is dropping a proposal directly into proposals/csharplang-10.0. In the past we dropped them into proposals and then migrated to the version specific folder close to release. In this particular case though the feature is approved, there is a PR to push it into the language so the intermediate step seems unneeded. Are we fine with this approach?

I'm fine with this.

> The PR is modifying an existing proposal in line.

I'm not fine with this. We should reference the portion of the original proposal that is being modified and update the text in the new proposal. The original proposal should be left as is.

File: proposals/csharp-7.0/pattern-matching.md
Comment (333fred) [inline proposals/csharp-7.0/pattern-matching.md:104]:
```suggestion
```

We shouldn't modify the original spec.

File: proposals/csharp-10.0/pattern-match-span-of-char-on-string.md
Comment (333fred) [reply proposals/csharp-10.0/pattern-match-span-of-char-on-string.md:23]: Yes, but I don't think that detail needs to be in the specification itself, or the motivation. I think it's fine to just have the first part of the sentence and leave out the optimize part.

File: proposals/csharp-10.0/pattern-match-span-of-char-on-string.md
Comment (jaredpar) [reply proposals/csharp-10.0/pattern-match-span-of-char-on-string.md:37]: Can we separate out this section. I think we should essentially keep the pattern matching part about what is legal in pattern matching. Hence keep this sentence. 

>  If *e* is of type `System.Span<char>` or `System.ReadOnlySpan<char>`, and *c* is a constant string, and *c* does not have a constant value of `null`, then the pattern is considered matching if `System.MemoryExtensions.SequenceEqual<char>(e, System.MemoryExtensions.AsSpan(c))` returns `true`

Then lets have a separate part where we discuss how these types are looked up and matched. I wouldn't us the `>` style for this. It's a part of this sec hence I'd put it here just in normal font. 

Believe we've done that in other places and I think it makes it easier to read.

**Review by jaredpar — COMMENTED** [review]
Looking good. Minor formatting feedback but should be good to go after that.

File: proposals/csharp-10.0/pattern-match-span-of-char-on-string.md
Comment (YairHalberstadt) [reply proposals/csharp-10.0/pattern-match-span-of-char-on-string.md:37]: Done

File: proposals/csharp-10.0/pattern-match-span-of-char-on-string.md
Comment (333fred) [reply proposals/csharp-10.0/pattern-match-span-of-char-on-string.md:20]: @YairHalberstadt I'm just waiting on this before signing off.

Comment (YairHalberstadt) [reply proposals/csharp-10.0/pattern-match-span-of-char-on-string.md:20]: @jaredpar seemed to want more info on the optimizations, so was waiting for a conclusion :-)

Comment (333fred) [reply proposals/csharp-10.0/pattern-match-span-of-char-on-string.md:20]: We don't have any info on optimizations, and I'd prefer to leave them out of the spec entirely.

Comment (YairHalberstadt) [reply proposals/csharp-10.0/pattern-match-span-of-char-on-string.md]: Change made

**Review by 333fred — APPROVED** [review]

**Comment by RikkiGibson** [comment]
@333fred @jaredpar is there anything blocking merge of this?

**Comment by 333fred** [comment]
I was waiting for @jaredpar to sign off, but I meant to ping him and accidentally had this drop off my radar.

**Review by jaredpar — APPROVED** [review]

**Comment by 333fred** [comment]
Thanks @YairHalberstadt

**Pull request merged**
